### PR TITLE
drivers and samples: Infineon CAT1 counter driver

### PR DESCRIPTION
**Pull request by pavloinfineon**
Added initial version of Infineon CAT1 counter driver
Added external trigger pin that is able to run counter
Added Pulse Counter sample

### CODE REVIEW

File: samples/boards/infineon/pulse_counter/sample.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+sample:
+  name: Pulse Counter Driver Test
```
Comment (ifyall) [inline samples/boards/infineon/pulse_counter/sample.yaml:2]:
```suggestion
  name: Pulse Counter Driver Example
```

File: drivers/counter/CMakeLists.txt [hunk 2]
```diff
@@ -37,8 +37,8 @@ zephyr_library_sources_ifdef(CONFIG_COUNTER_XLNX_AXI_TIMER      counter_xlnx_axi
 zephyr_library_sources_ifdef(CONFIG_COUNTER_TMR_ESP32           counter_esp32_tmr.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_RTC_ESP32           counter_esp32_rtc.c)
 zephyr_library_sources_ifdef(CONFIG_COUNTER_MICROCHIP_MCP7940N  rtc_mcp7940n.c)
-zephyr_library_sources_ifdef(CONFIG_COUNTER_ANDES_ATCPIT100     counter_andes_atcpit100.c)
```
Comment (ifyall) [inline drivers/counter/CMakeLists.txt:40]: Should not remove a non-Infineon driver.

**Review by ifyall — CHANGES_REQUESTED** [review]

**Review by ifyall — DISMISSED** [review]

File: drivers/counter/Kconfig.ifx_cat1 [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Infineon CAT1 counter driver
+
+# Copyright (c) 2022 Cypress Semiconductor Corporation (an Infineon company) or
```
Comment (fabiobaltieri) [inline drivers/counter/Kconfig.ifx_cat1:3]: 2023?

File: drivers/counter/counter_ifx_cat1.c [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+	cyhal_timer_cfg_t counter_cfg;
+	struct counter_alarm_cfg alarm_cfg_counter;
+	struct counter_top_cfg top_value_cfg_counter;
+	uint32_t guard_period;
+	cyhal_resource_inst_t hw_resource;
+	bool alarm_irq_flag;
+};
+
+/* Default Counter configuration structure */
+const cy_stc_tcpwm_counter_config_t _cyhal_timer_default_config = {
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:47]: Drop the `_` prefix, we don't use it in Zephyr. Also make it `static`.

File: drivers/counter/counter_ifx_cat1.c [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+	.reloadInput = CY_TCPWM_INPUT_0,
+	.startInputMode = 0x3U,
+	.startInput = CY_TCPWM_INPUT_0,
+	.stopInputMode = 0x3U,
+	.stopInput = CY_TCPWM_INPUT_0,
+	.countInputMode = 0x3U,
+	.countInput = CY_TCPWM_INPUT_1,
+};
+
+static int32_t _get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:69]: Drop the `_` prefixes.

File: drivers/counter/counter_ifx_cat1.c [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+{
+	int32_t ret = -1;
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:71]: How about using a more meaningful error code?

File: drivers/counter/counter_ifx_cat1.c [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+static void ifx_cat1_counter_set_int_pending(const struct device *dev)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+
+	cyhal_timer_enable_event(&data->counter_obj, CYHAL_TIMER_IRQ_CAPTURE_COMPARE,
+				 config->irq_priority, true);
+	NVIC_SetPendingIRQ((IRQn_Type)config->irqn);
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:136]: Would it make more sense to have `irqn` of the right type in the struct and only cast it once on the initial assignment?

File: drivers/counter/counter_ifx_cat1.c [hunk 8]
```diff
@@ -0,0 +1,488 @@
...
+static int ifx_cat1_counter_set_top_value(const struct device *dev,
+					  const struct counter_top_cfg *cfg)
+{
+	__ASSERT_NO_MSG(dev != NULL);
+	__ASSERT_NO_MSG(cfg != NULL);
+
+	cy_rslt_t rslt;
+	struct ifx_cat1_counter_data *const data = dev->data;
+	const struct ifx_cat1_counter_config *const config = dev->config;
+	bool ticks_gt_period = false;
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:262]: No need to initialize this one.

File: samples/boards/infineon/pulse_counter/src/main.c [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+
+void main(void)
+{
+	int ret;
+
+	printk("Pulse Counter Sample\n\n");
+
+	const struct device *p_dev_counter1 = COUNTER_1;
+
+	int err;
```
Comment (fabiobaltieri) [inline samples/boards/infineon/pulse_counter/src/main.c:70]: this looks a bit odd, do the declaration first and printk after?

File: samples/boards/infineon/pulse_counter/src/main.c [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+	if (!device_is_ready(p_dev_counter1)) {
+		printk("Counter 1 is not ready.\n");
+		return;
+	}
+
+	if (!device_is_ready(pulse.port)) {
```
Comment (fabiobaltieri) [inline samples/boards/infineon/pulse_counter/src/main.c:77]: You can use `gpio_is_ready_dt` here

File: samples/boards/infineon/pulse_counter/src/main.c [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+
+static void cb_alarm(const struct device *counter_dev, uint8_t chan_id, uint32_t ticks,
+		     void *user_data)
+{
+	int err;
+
+	printk("!! Alarm !!\n");
+	printk("!! Count = %u\n", ticks);
+
+	struct counter_alarm_cfg *p_alarm = (struct counter_alarm_cfg *)user_data;
```
Comment (fabiobaltieri) [inline samples/boards/infineon/pulse_counter/src/main.c:50]: Same as in main, keep these at the top, not sure it's explicitly in the coding style but that's how most of the code is written.

File: samples/boards/infineon/pulse_counter/src/main.c [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+#include <zephyr/sys/printk.h>
+
+struct counter_alarm_cfg alarm_cfg;
+
+#define COUNTER_1  DEVICE_DT_GET(DT_NODELABEL(counter1_0))
+#define PULSE_PIN  1
+#define PULSE_NODE DEVICE_DT_GET(DT_NODELABEL(gpio_prt9))
+
+struct counter_top_cfg counter_top;
+int counter_top_userdata;
```
Comment (fabiobaltieri) [inline samples/boards/infineon/pulse_counter/src/main.c:22]: static?

**Comment by pavloinfineon** [comment]
twister build 2 CI is failing compiling `drivers/counter/counter_ll_stm32_timer.c` which is unrelated to changes done in the recent commits. Double checked that kconfig or drivers.counter.basic_api changes look to be as expected so unsure what is causing these errors.
I was also unable to reproduce these compile errors locally.

File: drivers/counter/counter_ifx_cat1.c (hunk 4 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Updated

File: drivers/counter/counter_ifx_cat1.c (hunk 5 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Updated

File: drivers/counter/counter_ifx_cat1.c (hunk 6 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Changed to EINVAL

File: drivers/counter/counter_ifx_cat1.c (hunk 7 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Changed `config->irqn` to IRQn_TYPE

File: drivers/counter/counter_ifx_cat1.c (hunk 8 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Fixed

File: samples/boards/infineon/pulse_counter/src/main.c (hunk 9 above)
Comment (pavloinfineon) [reply samples/boards/infineon/pulse_counter/src/main.c]: Fixed order and moved printk

File: samples/boards/infineon/pulse_counter/src/main.c (hunk 10 above)
Comment (pavloinfineon) [reply samples/boards/infineon/pulse_counter/src/main.c]: Changed

File: samples/boards/infineon/pulse_counter/src/main.c (hunk 11 above)
Comment (pavloinfineon) [reply samples/boards/infineon/pulse_counter/src/main.c]: Fixed. Also changed in a couple other functions

File: samples/boards/infineon/pulse_counter/src/main.c (hunk 12 above)
Comment (pavloinfineon) [reply samples/boards/infineon/pulse_counter/src/main.c]: Changed

File: drivers/counter/counter_ifx_cat1.c [hunk 13]
```diff
@@ -0,0 +1,537 @@
...
+static int get_hw_block_info(TCPWM_CNT_Type *reg_addr, cyhal_resource_inst_t *hw_resource)
+{
+	int ret = -EINVAL;
+	uint32_t i;
+
+	for (i = 0u; i < _CYHAL_TCPWM_INSTANCES; i++) {
+		TCPWM_Type *base = _CYHAL_TCPWM_DATA[i].base;
+
+		if (((uint32_t)reg_addr > (uint32_t)base) &&
+		    ((uint32_t)reg_addr < (uint32_t)base + sizeof(TCPWM_Type))) {
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:78]: Hey I think that a cleaner way to do this would be to use `uintptr_t` and `POINTER_TO_UINT`, you could initialize `base` as:
```c
uintptr_t base = POINTER_TO_UINT(_CYHAL_TCPWM_DATA[i].base)
```

add a temporary variable for `reg_addr` and I guess `cnt`, and then you would not have to carry castings around the whole function.

File: drivers/counter/counter_ifx_cat1.c [hunk 14]
```diff
@@ -0,0 +1,537 @@
...
+
+	/* Decrement value to detect also case when val == counter_read(dev). Otherwise,
+	 * condition would need to include comparing diff against 0.
+	 */
+	uint32_t curr = cyhal_timer_read(&data->counter_obj);
+	uint32_t diff = ifx_cat1_counter_ticks_sub((val - 1), curr, top_val);
+
+	if ((absolute && (val < curr)) || (diff > max_rel_val)) {
+		if (absolute) {
+			err = -ETIME;
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:404]: The exit path of this function is a bit hard to follow, how about:
- change the function exit on 445 to just `return 0`
- move this after the `irq_on_late` check and do a `return -ETIME` directly (the function behavior is the same)
- drop `err`, you won't need it anymore

I think that would make it more obvious to understand what's going on.

**Comment by fabiobaltieri** [comment]
> Twister build 2 CI is failing compiling `drivers/counter/counter_ll_stm32_timer.c` which is unrelated to changes done in the recent commits.

Yeah it was broken upstream, should be fixed now, sorry about it, in this situation you could check on #ci in Discord or check recent PRs to see if someone is working on it or wait it out and force push again to trigger another CI run. (or send a fix :-))

File: drivers/counter/counter_ifx_cat1.c (hunk 14 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Changed in most recent push

File: drivers/counter/counter_ifx_cat1.c (hunk 13 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Simplified the casts around the function

**Review by gmarull — DISMISSED** [review]
Board specific samples should rarely exist when using driver APIs (the purpose of them is to be able to write portable applications, actually).

File: drivers/counter/counter_ifx_cat1.c [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+		if ((reg_addr_ptr > base) && (reg_addr_ptr < end_addr)) {
+
+			hw_resource->type = CYHAL_RSC_TCPWM;
+			hw_resource->block_num = i;
+			hw_resource->channel_num = ((reg_addr_ptr - cnt) / sizeof(TCPWM_CNT_Type));
+
+			if (hw_resource->channel_num < _CYHAL_TCPWM_DATA[i].num_channels) {
+				ret = 0;
+			}
+			break;
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:89]: The error path here is quite hard to follow, if I understand the intent could you change it to:
- `return 0` on like 87
- `return -EINVAL` on line 89 and 93

That should be easier to understand. Even better flip the condition on line 86 and make that the error path.

**Comment by pavloinfineon** [comment]
> Board specific samples should rarely exist when using driver APIs (the purpose of them is to be able to write portable applications, actually).

Removed our board specific sample

File: drivers/counter/counter_ifx_cat1.c (hunk 15 above)
Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: Flipped the conditions to make it easier to understand

File: drivers/counter/counter_ifx_cat1.c [hunk 16]
```diff
@@ -0,0 +1,537 @@
...
+#include <cyhal_timer.h>
+#include <cyhal_gpio.h>
+
+#include <cyhal_tcpwm_common.h>
+#include <zephyr/dt-bindings/pinctrl/ifx_cat1-pinctrl.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ifx_cat1_counter, CONFIG_COUNTER_LOG_LEVEL);
+
+/* Device config structure */
```
Comment (fabiobaltieri) [inline drivers/counter/counter_ifx_cat1.c:25]: nit: I would drop this and the ones in line 34 and 46, the struct names are meaningful enough, no need to repeat those in comments, up to you though

Comment (pavloinfineon) [reply drivers/counter/counter_ifx_cat1.c]: You were correct, comments did seem redundant. Removed them

**Review by fabiobaltieri — APPROVED** [review]

**Review by ifyall — APPROVED** [review]
